### PR TITLE
Support follower read

### PR DESCRIPTION
**Pull request by shiyuhang0**
### What problem does this PR solve? 
close https://github.com/pingcap/tispark/issues/2534

design doc also in this pr

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    val config = TiConfiguration.createDefault(pdAddresses)
+    ClientSession.getInstance(config).close()
+  }
+
+  test("follower read with invalid label should throw exception") {
+    judgeCancel
+    val spark_new = spark.newSession()
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read.label", "key1=value2")
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read", "follower")
+    assertThrows[Exception] {
```
Comment (xuanyu66) [inline core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala:75]: assert with specific exception type and message, such as
```
    the[SQLException] thrownBy {
      spark.sql(s"delete from $dbtable where i = 3")
    } should have message s"SELECT command denied to user $user@% for table $dbtable"
```

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

File: core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala (hunk 1 above)
Comment (shiyuhang0) [reply core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala:75]: The exception has been swallowed for the poor tikv-client implement
https://github.com/tikv/client-java/blob/1b5edcd8ab8ee12e4dbb84e7e4c008075d473a7a/src/main/java/org/tikv/common/region/RegionManager.java#L142

The Exception is `Can not get enough candidates` original, but now is npe which is meanless

Comment (xuanyu66) [reply core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala:75]: WTF -_-

**Review by xuanyu66 — APPROVED** [review]

File: docs/design/2022-09-16-follower-read.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+The Follower Read feature supports reading with follower or learner replica of a region under the premise of strongly consistent reads.
+
+TiSpark will support follower read after this doc.
+
+## Detailed design
+
+### New configuration
+TiSpark supports follower read with the following configurations
+- spark.tispark.replica_read：Read data from specified role. The optional roles are leader, follower and learner. You can also specify multiple roles, and we will pick the roles you specify in order. 
+- spark.tispark.replica_read.label：Only select TiKV store match specified labels. Format: label_x=value_x,label_y=value_y
```
Comment (Daemonxiao) [inline docs/design/2022-09-16-follower-read.md:22]: How to understand `replica_read.label`?

File: docs/features/follower_read.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# TiSpark Follower Read Feature
+
+The Follower Read feature supports reading with follower or learner replica of a region under the premise of strongly consistent reads.
```
Comment (zhangjinpeng1987) [inline docs/features/follower_read.md:3]: s/premise/promise

Comment (shiyuhang0) [reply docs/features/follower_read.md:3]: thank you!

**Review by Daemonxiao — APPROVED** [review]

**Pull request merged**